crates/networking/p2p/rlpx/connection/handshake.rs

```diff
@@ -21,80 +23,125 @@ use ethrex_rlp::{
     error::RLPDecodeError,
     structs::{Decoder, Encoder},
 };
+use futures::{StreamExt, stream::SplitStream};
 use k256::{
     PublicKey, SecretKey,
     ecdsa::{self, RecoveryId, SigningKey, VerifyingKey},
     elliptic_curve::sec1::ToEncodedPoint,
 };
 use rand::Rng;
 use sha3::{Digest, Keccak256};
-use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+use tokio::{
+    io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt},
+    net::{TcpSocket, TcpStream},
+    sync::Mutex,
+};
+use tokio_util::codec::Framed;
+
+use super::{
+    codec::RLPxCodec,
+    server::{Initiator, Receiver},
+};
 
 type Aes128Ctr64BE = ctr::Ctr64BE<aes::Aes128>;
 
 // https://github.com/ethereum/go-ethereum/blob/master/p2p/peer.go#L44
 pub const P2P_MAX_MESSAGE_SIZE: usize = 2048;
 
-pub(crate) async fn as_receiver<S>(
-    context: P2PContext,
-    peer_addr: SocketAddr,
-    mut stream: S,
-) -> Result<RLPxConnection<S>, RLPxError>
-where
-    S: AsyncRead + AsyncWrite + std::marker::Unpin,
-{
-    let remote_state = receive_auth(&context.signer, &mut stream).await?;
-    let local_state = send_ack(remote_state.public_key, &mut stream).await?;
-    let hashed_nonces: [u8; 32] =
-        Keccak256::digest([local_state.nonce.0, remote_state.nonce.0].concat()).into();
-    let node = Node::new(
-        peer_addr.ip(),
-        peer_addr.port(),
-        peer_addr.port(),
-        remote_state.public_key,
-    );
-    let codec = RLPxCodec::new(&local_state, &remote_state, hashed_nonces)?;
-    log_peer_debug(&node, "Completed handshake as receiver!");
-    Ok(RLPxConnection::new(
-        context.signer,
-        node,
-        stream,
-        codec,
-        context.storage,
-        context.blockchain,
-        context.client_version,
-        context.broadcast,
-    ))
+pub(crate) struct RemoteState {
+    pub(crate) public_key: H512,
+    pub(crate) nonce: H256,
+    pub(crate) ephemeral_key: PublicKey,
+    pub(crate) init_message: Vec<u8>,
 }
 
-pub(crate) async fn as_initiator<S>(
-    context: P2PContext,
-    node: Node,
-    mut stream: S,
-) -> Result<RLPxConnection<S>, RLPxError>
-where
-    S: AsyncRead + AsyncWrite + std::marker::Unpin,
-{
-    let local_state = send_auth(&context.signer, node.public_key, &mut stream).await?;
-    let remote_state = receive_ack(&context.signer, node.public_key, &mut stream).await?;
-    // Local node is initator
-    // keccak256(nonce || initiator-nonce)
-    let hashed_nonces: [u8; 32] =
-        Keccak256::digest([remote_state.nonce.0, local_state.nonce.0].concat()).into();
-    let codec = RLPxCodec::new(&local_state, &remote_state, hashed_nonces)?;
-    log_peer_debug(&node, "Completed handshake as initiator!");
-    Ok(RLPxConnection::new(
-        context.signer,
-        node,
+pub(crate) struct LocalState {
+    pub(crate) nonce: H256,
+    pub(crate) ephemeral_key: SecretKey,
+    pub(crate) init_message: Vec<u8>,
+}
+
+pub(crate) async fn perform(
```

Comment:
I find the name "perform" a bit confusing, can we add some comments on what this function does?

Reply:
Ah, I see, "perform" as in "perform handshake", a little bit confusing without context still

Reply:
Yeah... is to be used in the form `handshake::perform` but will think if there's a more expressive way